src/libraries/System.Private.CoreLib/src/System/Collections/Generic/List.cs

```diff
@@ -465,6 +465,38 @@ internal void Grow(int capacity)
             Capacity = newCapacity;
         }
 
+        /// <summary>
+        /// Enlarge this list so it may contain at least <paramref name="insertionCount"/> more elements
+        /// And copy data to their after-insertion positions.
+        /// This method is specifically for insertion, as it avoids 1 extra array copy.
+        /// You should only call this method when Count + insertionCount > Capacity.
+        /// </summary>
+        /// <param name="indexToInsert">Index of the first insertion.</param>
+        /// <param name="insertionCount">How many elements will be inserted.</param>
+        private void GrowForInsertion(int indexToInsert, int insertionCount = 1)
+        {
+            Debug.Assert(insertionCount > 0);
+
+            int capacity = checked(_size + insertionCount);
+
+            // Follow the same logic from Grow(capacity)
+
+            int newCapacity = _items.Length == 0 ? DefaultCapacity : 2 * _items.Length;
+            if ((uint)newCapacity > Array.MaxLength) newCapacity = Array.MaxLength;
+            if (newCapacity < capacity) newCapacity = capacity;
+
+            // Inline and adapt logic from set_Capacity
+
+            T[] newItems = new T[newCapacity];
+            if (indexToInsert != 0)
+                Array.Copy(_items, newItems, indexToInsert);
```

Comment:
nit: it would be nice to name the last argument to make it clear that it's length, not index (and a bug)

```suggestion
                Array.Copy(_items, newItems, length: indexToInsert);
```